### PR TITLE
Improve firmware flashing script with firmware download capability

### PR DESCRIPTION
**Pull request by Jo-Be-Co**
Added functionality to download firmware files from the Dresden Elektronik website when they are not available on the local instance (i.e. not included with the deCONZ binary).

### CODE REVIEW

**Comment by Jo-Be-Co** [comment]
This one could be further approved by reading in the possible firmware names. #261

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:76]: The script should check the return value of md5sum and exit if there is a sum mismatch with a printed error message.

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:42]: Spelling - "Alternatively"

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:44]: "or press Enter now to exit."
(No need to change this line.)

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:55]: "File not found locally. Try to download?"

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:95]: Are () or [] correct here to check retVal?

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:3]: Please bump the version to 0.6.

**Review by marthoc — CHANGES_REQUESTED** [review]
Thank you for these changes! Very useful feature that will be appreciated by many. Please make the initial changes that I have suggested (and for the other copies of the flashing script also) and I will review again and test.

**Comment by marthoc** [comment]
> This one could be further approved by reading in the possible firmware names. #261

I agree, but let's go with the small step first of allowing a user to specify a known firmware file before expanding functionality.

File: aarch64/root/firmware-update.sh
Comment (Jo-Be-Co) [reply aarch64/root/firmware-update.sh:95]: As far as I know:
1. Single `[ expr ]` will call the `test()` function
2. Double `[[ expr ]]` is a bash-internal.
3. Double `(( expr ))` is an arithmetic expression - `(( 9 > 10 ))` will evaluate differently than `[[ 9 > 10 ]]`. The latter checks for lexicographic order ...

As an addidtion: In arithmetic expression you don't need to use the `$` for numeric variables.

So as short answer: both work and are correct. The difference may be in comparing strings or numbers ...

File: amd64/root/firmware-update.sh
Comment (sorryusernameisalreadytaken) [inline amd64/root/firmware-update.sh:39]: I really want to be able to do a fast update of my raspbee-module. So I would love when I do not need to open my browser and check the latest firmware-file names.
```suggestion
echo "Remote firmware files available for flashing (limit to 10):"
echo " "

echo \$FW_ONLINE|tr \" \" \"\n\"
echo " "
```

File: aarch64/root/firmware-update.sh
Comment (sorryusernameisalreadytaken) [inline aarch64/root/firmware-update.sh:7]: Adding this for the preview-10-latest fw's and prefill (read) feature
```suggestion
FW_ONLINE=\$(curl -s \$FW_BASE?C=M;O=D'|sed -n 's/.*href=\"\([^\"]*\).*/\1/p'|grep -m 10 -E '^deCONZ_.*_0x.+[a-f0-9].bin.GCF\$')
FW_LATEST=\$(echo \$FW_ONLINE|tr \" \" \"\n\"|grep -m 1 GCF)
```

File: amd64/root/firmware-update.sh
Comment (sorryusernameisalreadytaken) [inline amd64/root/firmware-update.sh:21]: I would love to see a prefiled variable with the device path if there is just 1 result. This should be in most cases the result and speed up the process.

File: amd64/root/firmware-update.sh
Comment (sorryusernameisalreadytaken) [inline amd64/root/firmware-update.sh]: A prefilling is usefull.
```suggestion
read -e -p "File Name : " -i \$FW_LATEST fileName
```

**Comment by Jo-Be-Co** [comment]
> > This one could be further approved by reading in the possible firmware names. #261
> 
> I agree, but let's go with the small step first of allowing a user to specify a known firmware file before expanding functionality.

That's why I added this comment 😉

**Review by sorryusernameisalreadytaken — CHANGES_REQUESTED** [review]
I try to integrate some fct's from my 1-liner here:
https://github.com/marthoc/docker-deconz/issues/261#issuecomment-700556924
extende_raspbee_updater1liner

File: amd64/root/firmware-update.sh
Comment (Jo-Be-Co) [reply amd64/root/firmware-update.sh:39]: All the extra `\` aren't ment to be in your solution, are they?

File: aarch64/root/firmware-update.sh
Comment (Jo-Be-Co) [reply aarch64/root/firmware-update.sh:7]: I would put the list into an array. In that case the "latest" would be the first element in the array:
```shell
FW_LOCAL=( $(ls -1 "$FW_PATH") )
FW_ONLINE=( $(curl -s "$FW_BASE?C=M;O=D" | grep -Po '(?<=href=")deCONZ_[^"]+_0x[0-9a-f]+\.bin\.GCF(?=")') )
FW_LATEST=$FW_ONLINE

echo "additional files online (10-latest)"
printf "    %s\n" "${FW_ONLINE[@]}" | grep -m 10 -v "$(printf "%s\n" "${FW_LOCAL[@]}")"
```

@marthoc what do you think about the additional output? I wouldn't add the default `FW_LATEST` to the read as it won't fit the used hardware on most installations ...

File: amd64/root/firmware-update.sh
Comment (Jo-Be-Co) [reply amd64/root/firmware-update.sh]: But how do you make sure that the selection matches your hardware?

**Comment by marthoc** [comment]
@sorryusernameisalreadytaken @Jo-Be-Co I tend to agree with @Jo-Be-Co's comments above - I am all in favour of improving the usability of this tool, but making it faster or making assumptions for the user is not necessarily the way we get there. 

In fact, by making the tool "faster" we open ourselves (myself) up to a support headache when users wonder why the firmware file selected by the tool won't flash to their device, or why a listed firmware won't flash to their device, or why the tool is "recommending" certain firmwares. 

I like @Jo-Be-Co's initial approach (asking for the filename and attempting to download if it is not local) because in most cases where the listed firmwares aren't what the user wants, they will have been recommended a firmware during some type of support experience (i.e. try firmware X).

Another issue with presenting a list of firmwares to the user is that this list is generated by a lot of sed/grepping which can easily break if the output changes even slightly. This has happened before and is one of the reasons I abandoned this approach previously.

File: aarch64/root/firmware-update.sh
Comment (marthoc) [reply aarch64/root/firmware-update.sh]: Sounds correct to me. Thanks!

File: aarch64/root/firmware-update.sh
Comment (marthoc) [inline aarch64/root/firmware-update.sh:80]: Please revise this line to:
```
echo "Error comparing checksums! Please re-run this script..."
```

**Review by marthoc — CHANGES_REQUESTED** [review]

**Comment by marthoc** [comment]
You will also need to resolve the conflicts before we can merge.

**Comment by Jo-Be-Co** [comment]
> You will also need to resolve the conflicts before we can merge.

@marthoc Actually I'm a little lost on this as I am new to the fork thing in github. I do work on my own fork. With the github desktop client all of my commit go agains you repository.
But there I don't see any merge conflicts - do I need some kind of down stream merge?
Jogi

**Review by marthoc — APPROVED** [review]

**Comment by marthoc** [comment]
Thanks for this! Many will appreciate this feature.

**Comment by Jo-Be-Co** [comment]
> Thanks for this! Many will appreciate this feature.

@marthoc You're welcome! This is just a tiny bit a can give back in reply for this well done docker package 👍

Do you think it's worth it to investigate more into the two features that where mentioned here?

I could put these into seperate pull-requests or for the beginning into one or two issues ...

### List online Soruces
With the given code-snippets it would be straight forward to implement something like this:
```
Firmware available for flashing:
    deCONZ_ConBeeII_0x264a0700.bin.GCF
    deCONZ_ConBeeII_0x26580700.bin.GCF
    deCONZ_RaspBeeII_0x26610700.bin.GCF
    deCONZ_Rpi_0x26330500.bin.GCF
    deCONZ_Rpi_0x26350500.bin.GCF

Firmware available online for flashing (10 newest):
    deCONZ_ConBeeII_0x26420700.bin.GCF
    deCONZ_ConBeeII_0x26480700.bin.GCF
    deCONZ_ConBeeII_0x26490700.bin.GCF
    deCONZ_ConBeeII_0x26570700.bin.GCF
    deCONZ_ConBeeII_0x26650700.bin.GCF
    deCONZ_RaspBeeII_0x26650700.bin.GCF
    deCONZ_Rpi_0x26300500.bin.GCF
    deCONZ_Rpi_0x26320500.bin.GCF
    deCONZ_Rpi_0x26340500.bin.GCF
    deCONZ_Rpi_0x26370500.bin.GCF
```
### Evaluate device list
By evaluating the device list you could
1. check for valid input - or are there situations where the device path you have to enter is **not** listed by `GCFFlasher`?
1. use the device (some sort of) automatically if it is the only listed
1. use the type of the device from the list (in my case `ConBee II`) to filter valid firmware files local and/or online
### Bringing both together
The list of available firmware files could be starred or even filterd:
```
Firmware available for flashing:
  * deCONZ_ConBeeII_0x264a0700.bin.GCF
  * deCONZ_ConBeeII_0x26580700.bin.GCF
    deCONZ_RaspBeeII_0x26610700.bin.GCF
    deCONZ_Rpi_0x26330500.bin.GCF
    deCONZ_Rpi_0x26350500.bin.GCF

Firmware available online for flashing (10 newest):
  * deCONZ_ConBeeII_0x26420700.bin.GCF
  * deCONZ_ConBeeII_0x26480700.bin.GCF
  * deCONZ_ConBeeII_0x26490700.bin.GCF
  * deCONZ_ConBeeII_0x26570700.bin.GCF
 ** deCONZ_ConBeeII_0x26650700.bin.GCF
    deCONZ_RaspBeeII_0x26650700.bin.GCF
    deCONZ_Rpi_0x26300500.bin.GCF
    deCONZ_Rpi_0x26320500.bin.GCF
    deCONZ_Rpi_0x26340500.bin.GCF
    deCONZ_Rpi_0x26370500.bin.GCF
```

**Comment by marthoc** [comment]
@Jo-Be-Co I like the look of your suggestions, I'm just slightly worried that this is going to open up a problem with support (presenting too many options to users).

**Comment by Jo-Be-Co** [comment]
@Jo-Be-Co Just as you like! We could tailor it to any "size". I agree with you that the update process should be streamlined and just a single workflow.
So its more about assistance with informations and additional checks.

Anyhow - should we put this on hold, track it with an issue or shall I commit some lines for it?

**Pull request merged**
